File: crates/deepsec-scanner/src/scan.rs
use crate::detect::{DetectedTech, detect_tech, write_tech_json};
use crate::gate::evaluate_gate;
use crate::registry::MatcherRegistry;
use crate::matchers::NoiseTier;
use deepsec_core::ids::{generate_run_id, now_iso};
use deepsec_core::store::{file_hash_hex, read_file_record, write_file_record};
use deepsec_core::{
    CandidateMatch, DataRoot, FileRecord, FileStatus, RunPhase, RunType,
    ScannerConfig, ScannerMode, complete_run, create_run_meta, ensure_project, write_run_meta,
};
use globset::{Glob, GlobSetBuilder};
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone)]
pub struct ScanOptions {
    pub project_id: String,
    pub root: PathBuf,
    pub data_root: DataRoot,
    pub matcher_only: Vec<String>,
    pub matcher_exclude: Vec<String>,
    pub github_url: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct LanguageStat {
    pub language: String,
    pub files_scanned: usize,
    pub files_with_match: usize,
}

#[derive(Debug, Clone)]
pub struct ScanOutcome {
    pub run_id: String,
    pub files_scanned: usize,
    pub candidate_count: usize,
    pub detected: DetectedTech,
    pub active_matchers: Vec<String>,
    pub skipped_matchers: Vec<String>,
    pub language_stats: Vec<LanguageStat>,
}

#[derive(Debug, Clone)]
pub struct ScanFilesOutcome {
    pub run_id: String,
    pub files_scanned: usize,
    pub candidate_count: usize,
    pub detected: DetectedTech,
    pub active_matchers: Vec<String>,
    pub skipped_matchers: Vec<String>,
}

fn lang_for(rel: &Path) -> &'static str {
    match rel.extension().and_then(|e| e.to_str()) {
        Some("ts") | Some("tsx") => "typescript",
        Some("js") | Some("jsx") | Some("mjs") | Some("cjs") => "javascript",
        Some("py") => "python",
        Some("rb") => "ruby",
        Some("go") => "go",
        Some("rs") => "rust",
        Some("java") => "java",
        Some("kt") | Some("kts") => "kotlin",
        Some("cs") => "csharp",
        Some("php") => "php",
        Some("swift") => "swift",
        Some("ex") | Some("exs") => "elixir",
        Some("erl") => "erlang",
        Some("clj") | Some("cljs") => "clojure",
        Some("cr") => "crystal",
        Some("scala") => "scala",
        Some("c") | Some("h") => "c",
        Some("cpp") | Some("cc") | Some("hpp") => "cpp",
        Some("dart") => "dart",
        _ => "other",
    }
}

/// Run a full-repo scan: walk the project, evaluate matcher gates,
/// match against every file, write FileRecords, write RunMeta.
pub fn scan(opts: &ScanOptions) -> anyhow::Result<ScanOutcome> {
    ensure_project(
        &opts.data_root,
        &opts.project_id,
        opts.root.to_string_lossy().as_ref(),
        opts.github_url.as_deref(),
    )?;

    let run_id = generate_run_id();
    let mut meta = create_run_meta(
        &opts.project_id,
        &run_id,
        opts.root.to_string_lossy().as_ref(),
        RunType::Scan,
    );
    write_run_meta(&opts.data_root, &meta)?;

    let detected = detect_tech(&opts.root);
    write_tech_json(&opts.data_root, &opts.project_id, &detected)?;

    let mut registry = MatcherRegistry::with_builtin()?;
    registry.apply_filter(&opts.matcher_only, &opts.matcher_exclude);

    let mut active = Vec::new();
    let mut skipped = Vec::new();
    for m in registry.iter() {
        if evaluate_gate(m.requires(), &detected, &opts.root) {
            active.push(m.slug().to_string());
        } else {
            skipped.push(m.slug().to_string());
        }
    }

    let files = crate::walker::walk_project(&opts.root);
    let mut candidate_count = 0usize;
    let mut by_lang: BTreeMap<&'static str, LanguageStat> = BTreeMap::new();

    for rel in &files {
        let rel_str = rel.to_string_lossy().replace('\\', "/");
        let abs = opts.root.join(rel);
        let content = match fs_err::read_to_string(&abs) {
            Ok(s) => s,
            Err(_) => continue,
        };
        let normalized = content.replace("\r\n", "\n");

        let mut matches: Vec<CandidateMatch> = Vec::new();
        for m in registry.iter() {
            if !active.iter().any(|s| s == m.slug()) {
                continue;
            }
            if !matches_pattern(m.file_patterns(), &rel_str) {
                continue;
            }
            matches.extend(m.matches(&normalized, &rel_str));
        }

        let lang = lang_for(rel);
        let st = by_lang.entry(lang).or_insert_with(|| LanguageStat {
            language: lang.into(),
            ..Default::default()
        });
        st.files_scanned += 1;
        if !matches.is_empty() {
            st.files_with_match += 1;
        }
        candidate_count += matches.len();
        upsert_record(&opts.data_root, &opts.project_id, &rel_str, &run_id, &normalized, matches)?;
    }

    let total_files = files.len();
    meta.scanner_config = Some(ScannerConfig {
        matcher_slugs: active.clone(),
        mode: Some(ScannerMode::Full),
        source: None,
        file_count: None,
    });
    meta.stats.files_scanned = Some(total_files);
    meta.stats.candidates_found = Some(candidate_count);
    write_run_meta(&opts.data_root, &meta)?;
    complete_run(&opts.data_root, &opts.project_id, &run_id, RunPhase::Done)?;

    Ok(ScanOutcome {
        run_id,
        files_scanned: total_files,
        candidate_count,
        detected,
        active_matchers: active,
        skipped_matchers: skipped,
        language_stats: by_lang.into_values().collect(),
    })
}

/// Scan only the explicitly listed file paths. Writes a FileRecord for
/// every listed path, even when no matchers fire.
pub fn scan_files(opts: &ScanOptions, files: &[String], source: &str) -> anyhow::Result<ScanFilesOutcome> {
    ensure_project(
        &opts.data_root,
        &opts.project_id,
        opts.root.to_string_lossy().as_ref(),
        opts.github_url.as_deref(),
    )?;
    let run_id = generate_run_id();
    let mut meta = create_run_meta(
        &opts.project_id,
        &run_id,
        opts.root.to_string_lossy().as_ref(),
        RunType::Scan,
    );
    write_run_meta(&opts.data_root, &meta)?;

    let detected = detect_tech(&opts.root);
    write_tech_json(&opts.data_root, &opts.project_id, &detected)?;

    let mut registry = MatcherRegistry::with_builtin()?;
    registry.apply_filter(&opts.matcher_only, &opts.matcher_exclude);

    let mut active = Vec::new();
    let mut skipped = Vec::new();
    for m in registry.iter() {
        if evaluate_gate(m.requires(), &detected, &opts.root) {
            active.push(m.slug().to_string());
        } else {
            skipped.push(m.slug().to_string());
        }
    }

    let mut candidate_count = 0usize;
    for rel_str in files {
        let rel = Path::new(rel_str);
        let abs = opts.root.join(rel);
        let content = match fs_err::read_to_string(&abs) {
            Ok(s) => s,
            Err(_) => String::new(),
        };
        let normalized = content.replace("\r\n", "\n");
        let mut matches: Vec<CandidateMatch> = Vec::new();
        for m in registry.iter() {
            if !active.iter().any(|s| s == m.slug()) {
                continue;
            }
            if !matches_pattern(m.file_patterns(), rel_str) {
                continue;
            }
            matches.extend(m.matches(&normalized, rel_str));
        }
        candidate_count += matches.len();
        upsert_record(&opts.data_root, &opts.project_id, rel_str, &run_id, &normalized, matches)?;
    }

    meta.scanner_config = Some(ScannerConfig {
        matcher_slugs: active.clone(),
        mode: Some(ScannerMode::Files),
        source: Some(source.into()),
        file_count: Some(files.len()),
    });
    meta.stats.files_scanned = Some(files.len());
    meta.stats.candidates_found = Some(candidate_count);
    write_run_meta(&opts.data_root, &meta)?;
    complete_run(&opts.data_root, &opts.project_id, &run_id, RunPhase::Done)?;

    Ok(ScanFilesOutcome {
        run_id,
        files_scanned: files.len(),
        candidate_count,
        detected,
        active_matchers: active,
        skipped_matchers: skipped,
    })
}

fn matches_pattern(patterns: &[String], rel: &str) -> bool {
    let mut builder = GlobSetBuilder::new();
    let mut any = false;
    for p in patterns {
        if let Ok(g) = Glob::new(p) {
            builder.add(g);
            any = true;
        }
    }
    if !any {
        return false;
    }
    let Ok(set) = builder.build() else {
        return false;
    };
    set.is_match(rel)
}

fn upsert_record(
    root: &DataRoot,
    project_id: &str,
    rel: &str,
    run_id: &str,
    content: &str,
    mut new_matches: Vec<CandidateMatch>,
) -> anyhow::Result<()> {
    let hash = file_hash_hex(content.as_bytes());
    let now = now_iso();
    let existing = read_file_record(root, project_id, rel).ok().flatten();

    let mut rec = existing.unwrap_or(FileRecord {
        file_path: rel.into(),
        project_id: project_id.into(),
        candidates: Vec::new(),
        last_scanned_at: now.clone(),
        last_scanned_run_id: run_id.into(),
        file_hash: hash.clone(),
        findings: Vec::new(),
        analysis_history: Vec::new(),
        git_info: None,
        status: FileStatus::Pending,
        locked_by_run_id: None,
        locked_at: None,
    });

    rec.last_scanned_at = now;
    rec.last_scanned_run_id = run_id.into();
    rec.file_hash = hash;

    // dedupe new + existing candidates by (slug, pattern, line numbers)
    let mut seen: std::collections::HashSet<String> = rec
        .candidates
        .iter()
        .map(|c| format!("{}|{}|{:?}", c.vuln_slug, c.matched_pattern, c.line_numbers))
        .collect();
    new_matches.retain(|c| seen.insert(format!("{}|{}|{:?}", c.vuln_slug, c.matched_pattern, c.line_numbers)));
    rec.candidates.extend(new_matches);

    // If this file is pending and we found new candidates, it stays pending.
    // Don't downgrade `analyzed` files just because a re-scan ran.
    if matches!(rec.status, FileStatus::Pending) && rec.candidates.is_empty() {
        // no candidates → nothing to investigate yet
    }

    write_file_record(root, &rec)?;
    Ok(())
}

/// Group FileRecords into batches by directory, splitting oversized
/// directories. Matches the TS implementation's behavior.
pub fn batch_candidates(records: &[FileRecord], max_batch: usize) -> Vec<Vec<&FileRecord>> {
    let mut by_dir: BTreeMap<String, Vec<&FileRecord>> = BTreeMap::new();
    for r in records {
        let dir = std::path::Path::new(&r.file_path)
            .parent()
            .and_then(|p| p.to_str())
            .unwrap_or("")
            .to_string();
        by_dir.entry(dir).or_default().push(r);
    }
    let mut out = Vec::new();
    let mut current: Vec<&FileRecord> = Vec::new();
    for (_dir, files) in by_dir {
        for chunk in files.chunks(max_batch) {
            if current.len() + chunk.len() <= max_batch {
                current.extend(chunk);
            } else {
                if !current.is_empty() {
                    out.push(std::mem::take(&mut current));
                }
                out.push(chunk.to_vec());
            }
        }
    }
    if !current.is_empty() {
        out.push(current);
    }
    out
}

/// Score a file by its highest-precision matcher hit. Lower = higher priority.
pub fn noise_score(rec: &FileRecord, registry: &MatcherRegistry) -> u32 {
    let mut best: u32 = NoiseTier::Noisy.rank() + 1;
    for c in &rec.candidates {
        let t = registry.noise_tier(&c.vuln_slug);
        if t.rank() < best {
            best = t.rank();
        }
    }
    best
}

File: internal/scanner/scan.go
package scanner

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/bmatcuk/doublestar/v4"
	"github.com/noeljackson/deepsec/internal/core"
	scannerast "github.com/noeljackson/deepsec/internal/scanner/ast"
)

// Options bundles the parameters every scan call needs.
type Options struct {
	ProjectID      string
	Root           string
	DataRoot       core.DataRoot
	MatcherOnly    []string
	MatcherExclude []string
	// ExtraMatcherPaths is a list of TOML files (or directories of TOML files)
	// to load on top of the bundled matcher pack. User matchers registered
	// with a slug that already exists in the bundle override it.
	ExtraMatcherPaths []string
	GithubURL         string
	// ForceRescan disables the file-hash cache. When true, every file is
	// re-evaluated even when both its content and the matcher pack are
	// unchanged since the last scan.
	ForceRescan bool
}

// LanguageStat is one row of the per-language scan summary.
type LanguageStat struct {
	Language       string
	FilesScanned   int
	FilesWithMatch int
}

// Outcome is what `Scan` returns.
type Outcome struct {
	RunID           string
	FilesScanned    int
	CandidateCount  int
	CacheHits       int
	Detected        DetectedTech
	ActiveMatchers  []string
	SkippedMatchers []string
	LanguageStats   []LanguageStat
}

// FilesOutcome is what `ScanFiles` returns (no per-language stats, since
// the file list is bounded by the caller).
type FilesOutcome struct {
	RunID           string
	FilesScanned    int
	CandidateCount  int
	CacheHits       int
	Detected        DetectedTech
	ActiveMatchers  []string
	SkippedMatchers []string
}

// Scan walks every file under opts.Root and writes FileRecords with
// candidates merged in.
func Scan(opts Options) (*Outcome, error) {
	if _, err := opts.DataRoot.EnsureProject(opts.ProjectID, opts.Root, opts.GithubURL); err != nil {
		return nil, err
	}
	runID := core.GenerateRunID()
	meta := core.NewRunMeta(opts.ProjectID, runID, opts.Root, core.RunTypeScan)
	if err := opts.DataRoot.WriteRunMeta(meta); err != nil {
		return nil, err
	}

	tech := Detect(opts.Root)
	if err := WriteTechJSON(opts.DataRoot, opts.ProjectID, tech); err != nil {
		return nil, err
	}

	reg, err := loadRegistry(opts)
	if err != nil {
		return nil, err
	}
	active, skipped := splitGated(reg, tech, opts.Root)
	astRT, err := runtimeForActiveAST(active, reg)
	if err != nil {
		return nil, err
	}
	if astRT != nil {
		defer astRT.Close(context.Background())
	}

	files, err := WalkProject(opts.Root)
	if err != nil {
		return nil, err
	}

	packHash := reg.PackHash()
	candidateCount := 0
	cacheHits := 0
	byLang := map[string]*LanguageStat{}
	for _, rel := range files {
		abs := filepath.Join(opts.Root, rel)
		body, err := os.ReadFile(abs)
		if err != nil {
			continue
		}
		content := strings.ReplaceAll(string(body), "\r\n", "\n")
		fileHash := core.FileHashHex([]byte(content))
		lang := langFor(rel)
		st, ok := byLang[lang]
		if !ok {
			st = &LanguageStat{Language: lang}
			byLang[lang] = st
		}
		st.FilesScanned++
		cached, hit, err := cachedCandidates(opts, rel, fileHash, packHash)
		if err != nil {
			return nil, err
		}
		var matches []core.CandidateMatch
		if hit {
			matches = cached
			cacheHits++
		} else {
			matches = runMatchers(reg, active, content, rel, astRT)
		}
		if len(matches) > 0 {
			st.FilesWithMatch++
		}
		candidateCount += len(matches)
		if hit {
			if err := touchRecord(opts.DataRoot, opts.ProjectID, rel, runID, fileHash, packHash); err != nil {
				return nil, err
			}
			continue
		}
		if err := upsertRecord(opts.DataRoot, opts.ProjectID, rel, runID, content, packHash, matches); err != nil {
			return nil, err
		}
	}

	mode := core.ScannerModeFull
	meta.ScannerConfig = &core.ScannerConfig{
		MatcherSlugs: active,
		Mode:         mode,
	}
	fs := len(files)
	meta.Stats.FilesScanned = &fs
	meta.Stats.CandidatesFound = &candidateCount
	if err := opts.DataRoot.WriteRunMeta(meta); err != nil {
		return nil, err
	}
	if _, err := opts.DataRoot.CompleteRun(opts.ProjectID, runID, core.RunPhaseDone); err != nil {
		return nil, err
	}

	stats := make([]LanguageStat, 0, len(byLang))
	for _, s := range byLang {
		stats = append(stats, *s)
	}
	sort.Slice(stats, func(i, j int) bool { return stats[i].Language < stats[j].Language })

	return &Outcome{
		RunID:           runID,
		FilesScanned:    fs,
		CandidateCount:  candidateCount,
		CacheHits:       cacheHits,
		Detected:        tech,
		ActiveMatchers:  active,
		SkippedMatchers: skipped,
		LanguageStats:   stats,
	}, nil
}

// ScanFiles scans only the given relative file paths. Writes a
// FileRecord for every listed path, even when no matchers fire.
func ScanFiles(opts Options, files []string, source string) (*FilesOutcome, error) {
	if _, err := opts.DataRoot.EnsureProject(opts.ProjectID, opts.Root, opts.GithubURL); err != nil {
		return nil, err
	}
	runID := core.GenerateRunID()
	meta := core.NewRunMeta(opts.ProjectID, runID, opts.Root, core.RunTypeScan)
	if err := opts.DataRoot.WriteRunMeta(meta); err != nil {
		return nil, err
	}

	tech := Detect(opts.Root)
	if err := WriteTechJSON(opts.DataRoot, opts.ProjectID, tech); err != nil {
		return nil, err
	}

	reg, err := loadRegistry(opts)
	if err != nil {
		return nil, err
	}
	active, skipped := splitGated(reg, tech, opts.Root)
	astRT, err := runtimeForActiveAST(active, reg)
	if err != nil {
		return nil, err
	}
	if astRT != nil {
		defer astRT.Close(context.Background())
	}

	packHash := reg.PackHash()
	candidateCount := 0
	cacheHits := 0
	for _, rel := range files {
		abs := filepath.Join(opts.Root, rel)
		body, _ := os.ReadFile(abs)
		content := strings.ReplaceAll(string(body), "\r\n", "\n")
		fileHash := core.FileHashHex([]byte(content))
		cached, hit, err := cachedCandidates(opts, rel, fileHash, packHash)
		if err != nil {
			return nil, err
		}
		var matches []core.CandidateMatch
		if hit {
			matches = cached
			cacheHits++
		} else {
			matches = runMatchers(reg, active, content, rel, astRT)
		}
		candidateCount += len(matches)
		if hit {
			if err := touchRecord(opts.DataRoot, opts.ProjectID, rel, runID, fileHash, packHash); err != nil {
				return nil, err
			}
			continue
		}
		if err := upsertRecord(opts.DataRoot, opts.ProjectID, rel, runID, content, packHash, matches); err != nil {
			return nil, err
		}
	}

	count := len(files)
	mode := core.ScannerModeFiles
	meta.ScannerConfig = &core.ScannerConfig{
		MatcherSlugs: active,
		Mode:         mode,
		Source:       source,
		FileCount:    &count,
	}
	meta.Stats.FilesScanned = &count
	meta.Stats.CandidatesFound = &candidateCount
	if err := opts.DataRoot.WriteRunMeta(meta); err != nil {
		return nil, err
	}
	if _, err := opts.DataRoot.CompleteRun(opts.ProjectID, runID, core.RunPhaseDone); err != nil {
		return nil, err
	}

	return &FilesOutcome{
		RunID:           runID,
		FilesScanned:    count,
		CandidateCount:  candidateCount,
		CacheHits:       cacheHits,
		Detected:        tech,
		ActiveMatchers:  active,
		SkippedMatchers: skipped,
	}, nil
}

// loadRegistry builds the registry used by Scan / ScanFiles: bundled
// pack first, then any user-supplied paths from opts.ExtraMatcherPaths,
// then the only/exclude filter. Entries from extras override bundled
// entries that share a slug.
func loadRegistry(opts Options) (*Registry, error) {
	reg, err := WithBuiltin()
	if err != nil {
		return nil, err
	}
	for _, p := range opts.ExtraMatcherPaths {
		info, err := os.Stat(p)
		if err != nil {
			return nil, fmt.Errorf("matchers.extra_paths %q: %w", p, err)
		}
		if info.IsDir() {
			if err := reg.LoadTOMLDir(p); err != nil {
				return nil, fmt.Errorf("matchers.extra_paths %q: %w", p, err)
			}
			continue
		}
		if err := reg.LoadTOMLFile(p); err != nil {
			return nil, fmt.Errorf("matchers.extra_paths %q: %w", p, err)
		}
	}
	reg.ApplyFilter(opts.MatcherOnly, opts.MatcherExclude)
	return reg, nil
}

func splitGated(reg *Registry, tech DetectedTech, root string) (active, skipped []string) {
	for _, m := range reg.All() {
		if EvaluateGate(m.Requires(), tech, root) {
			active = append(active, m.Slug())
		} else {
			skipped = append(skipped, m.Slug())
		}
	}
	return
}

func runtimeForActiveAST(active []string, reg *Registry) (*scannerast.Runtime, error) {
	if !reg.HasASTPatterns() {
		return nil, nil
	}
	activeSet := map[string]struct{}{}
	for _, slug := range active {
		activeSet[slug] = struct{}{}
	}
	for _, m := range reg.All() {
		if _, ok := activeSet[m.Slug()]; ok && m.HasASTPatterns() {
			return scannerast.NewRuntime(context.Background(), scannerast.DefaultGrammars())
		}
	}
	return nil, nil
}

func runMatchers(reg *Registry, active []string, content, rel string, astRT *scannerast.Runtime) []core.CandidateMatch {
	out := make([]core.CandidateMatch, 0)
	activeSet := map[string]struct{}{}
	for _, s := range active {
		activeSet[s] = struct{}{}
	}
	astLang := scannerast.LanguageForPath(rel)
	type astPlan struct {
		matcher  *Matcher
		patterns []compiledASTPattern
	}
	var astPlans []astPlan
	for _, m := range reg.All() {
		if _, ok := activeSet[m.Slug()]; !ok {
			continue
		}
		if !matchesAnyGlob(m.FilePatterns(), rel) {
			continue
		}
		if astRT == nil || astLang == "" || !m.HasASTLanguage(astLang) {
			out = append(out, m.Match(content, rel)...)
			continue
		}
		patterns := m.EligibleASTPatterns(content, rel, astLang)
		if len(patterns) > 0 {
			astPlans = append(astPlans, astPlan{matcher: m, patterns: patterns})
		}
	}
	if len(astPlans) == 0 {
		return dedupeCandidates(out)
	}
	tree, err := astRT.Parse(context.Background(), astLang, []byte(content), rel)
	if err != nil {
		if errors.Is(err, scannerast.ErrLanguageUnavailable) {
			return dedupeCandidates(out)
		}
		return dedupeCandidates(out)
	}
	defer tree.Close()
	for _, plan := range astPlans {
		out = append(out, plan.matcher.MatchAST(tree, rel, plan.patterns)...)
	}
	return dedupeCandidates(out)
}

func dedupeCandidates(in []core.CandidateMatch) []core.CandidateMatch {
	if len(in) < 2 {
		return in
	}
	seen := map[string]struct{}{}
	out := make([]core.CandidateMatch, 0, len(in))
	for _, c := range in {
		key := candidateKey(c)
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		out = append(out, c)
	}
	return out
}

func matchesAnyGlob(patterns []string, rel string) bool {
	for _, p := range patterns {
		ok, _ := doublestar.PathMatch(p, rel)
		if ok {
			return true
		}
	}
	return false
}

func upsertRecord(root core.DataRoot, projectID, rel, runID, content, matcherPackHash string, newMatches []core.CandidateMatch) error {
	now := core.NowISO()
	hash := core.FileHashHex([]byte(content))
	existing, err := root.ReadFileRecord(projectID, rel)
	if err != nil {
		return err
	}
	if existing == nil {
		existing = &core.FileRecord{
			FilePath:            rel,
			ProjectID:           projectID,
			Candidates:          []core.CandidateMatch{},
			LastScannedAt:       now,
			LastScannedRunID:    runID,
			FileHash:            hash,
			LastMatcherPackHash: matcherPackHash,
			Findings:            []core.Finding{},
			AnalysisHistory:     []core.AnalysisEntry{},
			Status:              core.StatusPending,
		}
	}
	existing.LastScannedAt = now
	existing.LastScannedRunID = runID
	existing.FileHash = hash
	existing.LastMatcherPackHash = matcherPackHash

	seen := map[string]struct{}{}
	for _, c := range existing.Candidates {
		seen[candidateKey(c)] = struct{}{}
	}
	for _, c := range newMatches {
		if _, ok := seen[candidateKey(c)]; ok {
			continue
		}
		seen[candidateKey(c)] = struct{}{}
		existing.Candidates = append(existing.Candidates, c)
	}
	return root.WriteFileRecord(existing)
}

// cachedCandidates returns this file's existing candidates if the file
// content hash AND the matcher pack hash both match the previous scan.
// `hit` is true exactly when the caller should skip matcher execution.
// `--force-rescan` (opts.ForceRescan) always returns hit=false.
func cachedCandidates(opts Options, rel, fileHash, packHash string) ([]core.CandidateMatch, bool, error) {
	if opts.ForceRescan {
		return nil, false, nil
	}
	existing, err := opts.DataRoot.ReadFileRecord(opts.ProjectID, rel)
	if err != nil {
		return nil, false, err
	}
	if existing == nil {
		return nil, false, nil
	}
	if existing.FileHash != fileHash {
		return nil, false, nil
	}
	if existing.LastMatcherPackHash != packHash {
		return nil, false, nil
	}
	return existing.Candidates, true, nil
}

// touchRecord updates LastScannedAt/RunID/FileHash/LastMatcherPackHash
// for a cache hit without rewriting Candidates or other fields.
func touchRecord(root core.DataRoot, projectID, rel, runID, fileHash, packHash string) error {
	existing, err := root.ReadFileRecord(projectID, rel)
	if err != nil {
		return err
	}
	if existing == nil {
		// Should not happen for a cache hit, but be defensive.
		return nil
	}
	existing.LastScannedAt = core.NowISO()
	existing.LastScannedRunID = runID
	existing.FileHash = fileHash
	existing.LastMatcherPackHash = packHash
	return root.WriteFileRecord(existing)
}

func candidateKey(c core.CandidateMatch) string {
	var lines []string
	for _, l := range c.LineNumbers {
		lines = append(lines, intToString(l))
	}
	return c.VulnSlug + "|" + c.MatchedPattern + "|" + strings.Join(lines, ",")
}

func intToString(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}

func langFor(rel string) string {
	switch strings.ToLower(filepath.Ext(rel)) {
	case ".ts", ".tsx":
		return "typescript"
	case ".js", ".jsx", ".mjs", ".cjs":
		return "javascript"
	case ".py":
		return "python"
	case ".rb":
		return "ruby"
	case ".go":
		return "go"
	case ".rs":
		return "rust"
	case ".java":
		return "java"
	case ".kt", ".kts":
		return "kotlin"
	case ".cs":
		return "csharp"
	case ".php":
		return "php"
	case ".swift":
		return "swift"
	case ".ex", ".exs":
		return "elixir"
	case ".erl":
		return "erlang"
	case ".clj", ".cljs":
		return "clojure"
	case ".cr":
		return "crystal"
	case ".scala":
		return "scala"
	case ".c", ".h":
		return "c"
	case ".cpp", ".cc", ".hpp":
		return "cpp"
	case ".dart":
		return "dart"
	}
	return "other"
}

// BatchRecords groups records into directory-coherent batches, splitting
// oversized directories. Mirrors the Rust port's heuristic.
func BatchRecords(records []*core.FileRecord, maxBatch int) [][]*core.FileRecord {
	if maxBatch < 1 {
		maxBatch = 1
	}
	byDir := map[string][]*core.FileRecord{}
	dirs := []string{}
	for _, r := range records {
		d := filepath.Dir(r.FilePath)
		if _, ok := byDir[d]; !ok {
			dirs = append(dirs, d)
		}
		byDir[d] = append(byDir[d], r)
	}
	sort.Strings(dirs)
	out := [][]*core.FileRecord{}
	current := []*core.FileRecord{}
	for _, d := range dirs {
		files := byDir[d]
		for i := 0; i < len(files); i += maxBatch {
			end := i + maxBatch
			if end > len(files) {
				end = len(files)
			}
			chunk := files[i:end]
			if len(current)+len(chunk) <= maxBatch {
				current = append(current, chunk...)
			} else {
				if len(current) > 0 {
					out = append(out, current)
					current = nil
				}
				out = append(out, append([]*core.FileRecord(nil), chunk...))
			}
		}
	}
	if len(current) > 0 {
		out = append(out, current)
	}
	return out
}
